### ISSUE TITLE
add_port("ez", impedance=50) without extent= silently fails to couple to patch antennas

### ISSUE THREAD

**Issue by bk-squared** [bk-squared/rfx#71]
## Summary

`sim.add_port((x, y, z), \"ez\", impedance=R)` without an `extent=` kwarg creates a single-cell LumpedPort. When the port is placed mid-substrate between a ground plane and a microstrip patch (standard edge-fed inset geometry), the resulting \"port\" is a floating single-cell Ez source inside dielectric, decoupled from the patch TM mode. The simulation runs, the optimizer reports a loss (via `late_energy / early_energy` proxies), but the true |S11|(f) is ~0 dB across the band — the antenna is not excited.

## Reproduction

Minimal CPU smoke (validated 2026-04-18, Ubuntu 22.04, JAX-CPU):

```python
import numpy as np
from rfx import Simulation, Box, GaussianPulse

F0, EPS_R, H_SUB, DX = 10e9, 3.38, 0.787e-3, 0.5e-3
W0 = 2.998e8 / (2 * F0) * np.sqrt(2 / (EPS_R + 1))
L0 = 2.998e8 / (2 * F0 * np.sqrt((EPS_R + 1) / 2))  # uncorrected
CPML = 8
z_offset = (CPML + 2) * DX
patch_z = z_offset + DX + H_SUB
dom = (40e-3, 30e-3, z_offset + DX + H_SUB + 30e-3 + CPML * DX)

def build(fix):
    sim = Simulation(freq_max=2*F0, domain=dom, boundary=\"cpml\",
                     cpml_layers=CPML, dx=DX)
    sim.add_material(\"sub\", eps_r=EPS_R)
    sim.add(Box((8e-3, 8e-3, z_offset), (dom[0]-8e-3, dom[1]-8e-3, z_offset+DX)), material=\"pec\")
    sim.add(Box((8e-3, 8e-3, z_offset+DX), (dom[0]-8e-3, dom[1]-8e-3, z_offset+DX+H_SUB)), material=\"sub\")
    xc, yc = dom[0]/2, dom[1]/2
    sim.add(Box((xc-L0/2, yc-W0/2, patch_z), (xc+L0/2, yc+W0/2, patch_z+DX)), material=\"pec\")
    fx = xc - L0/2 + L0/3     # edge-inset feed
    if fix:  # with extent -> WirePort spans the substrate
        sim.add_port((fx, yc, z_offset+DX), \"ez\", impedance=50.0,
                     waveform=GaussianPulse(f0=F0, bandwidth=0.5), extent=H_SUB)
    else:    # without extent -> single-cell LumpedPort at mid-substrate
        sim.add_port((fx, yc, z_offset+DX+H_SUB/2), \"ez\", impedance=50.0,
                     waveform=GaussianPulse(f0=F0, bandwidth=0.5))
    sim.add_probe((fx, yc, z_offset+DX+H_SUB/2), \"ez\")
    return sim

freqs = np.linspace(5e9, 15e9, 101)
for fix in (False, True):
    sim = build(fix); g = sim._build_grid()
    r = sim.run(n_steps=g.num_timesteps(num_periods=30),
                compute_s_params=True, s_param_freqs=freqs)
    s11 = np.asarray(r.s_params)[0, 0, :]
    print(f\"fix={fix}: |S11|_min = {20*np.log10(np.abs(s11)).min():.2f} dB at {freqs[np.argmin(np.abs(s11))]/1e9:.2f} GHz\")
```

Observed:
```
fix=False (no extent): |S11|_min = -1.66 dB at 15.00 GHz
fix=True  (extent=H_SUB): |S11|_min = -17.87 dB at 14.30 GHz
```

## Impact

In a production inverse-design workflow (4-element series-fed patch array, 14 scalar design vars, Adam 30 iter, loss = 0.3·S11_proxy + 0.7·directivity_ratio):

- With `extent=` missing: optimiser reports loss = -0.40, peak θ = 29.4°, D(30°) > D(0°) — *appears* successful.
- Actual |S11| < -1 dB across 5-15 GHz — antenna does not radiate.
- The reported beam is driven by the tiny fraction of power that leaks past the broken port into NTFF DFT. Directivity ratio maximises against numerical noise.
- Optimiser is \"gaming\" a broken measurement.

Full trace / plots in our paper's reproducer: `papers/rfx-tap/experiments/ex3_series_fed_array.py` + `results/ex3/s11.png` (flat line near 0 dB).

## Why it happens

- No `extent=` → single-cell LumpedPort (`rfx/sources/sources.py:186-244`).
- Port position mid-substrate → the Ez cell is sandwiched between dielectric above and dielectric below.
- No conducting pin or microstrip trace connects the port cell to the patch.
- Patch's TM10 mode requires a vertical current (coax probe) or a surface current (microstrip feed) coupling. Neither is present.

The `WirePort` branch (`rfx/sources/sources.py:247-309`) distributes the Ez source over multiple cells along the feed axis, effectively emulating a coax pin. Use `extent=H_SUB` (substrate height) for patch antennas fed from ground → patch plane.

## Suggested fixes (pick any)

1. **Default `extent=H_SUB` when port type is \"ez\" and port position is inside a substrate-like geometry** (detectable via material probe at port cell). Auto-upgrade to WirePort with a UserWarning.
2. **Preflight warning**: when a single-cell LumpedPort is registered inside a dielectric with no adjacent PEC pin / microstrip, emit a warning explaining that the port may not couple to antenna modes. Point the user at `extent=`.
3. **Documentation**: add a \"common pitfalls\" note to `add_port` docstring and to the `examples/crossval/05_patch_antenna.py` reference highlighting this. The ex1-DRA example in the paper (`papers/rfx-tap/experiments/ex1_dra.py:177-183`) has the correct `extent=DRA_Z/3` pattern — make it canonical.

Option (2) alone would have saved several days of debugging in our workflow.

## Environment

- rfx: `main` branch @ 2026-04-17
- JAX: nvcr.io/nvidia/jax:24.10-py3
- Reproduced uniform and NU grids.

## Related

- See companion issue about `run_nonuniform` + `extent=`-based WirePort + `jax.grad` crashing with TracerArrayConversionError.

**Comment by bk-squared** [comment]
Fixed in 7d41d8f: added P1.9 preflight check in `rfx/api.py` that emits a `UserWarning` when a single-cell `LumpedPort` sits inside a dielectric (eps_r > 1.001) with no PEC geometry within ±dx·1.01 along the port's component axis. Warning cites `extent=<substrate_height>` as the mitigation. Under `preflight(strict=True)` it escalates to `ValueError`. Gated on `impedance > 0 AND extent is None AND excite is True` so `add_source` (impedance=0) and WirePort callers are unaffected. Tests: `tests/test_port_preflight.py` (5 cases: positive + 3 negatives + strict).
